test/transcoding/fcmp.ll

```diff
@@ -0,0 +1,402 @@
+; CHECK-SPIRV: 5 FUnordLessThanEqual [[#bool]] [[#r76]]
+; CHECK-SPIRV: 5 FUnordLessThanEqual [[#bool]] [[#r77]]
+; CHECK-SPIRV: 5 FUnordLessThanEqual [[#bool]] [[#r78]]
+; CHECK-SPIRV: 5 FUnordLessThanEqual [[#bool]] [[#r79]]
+; CHECK-SPIRV: 5 FUnordLessThanEqual [[#bool]] [[#r80]]
+; CHECK-SPIRV: 5 FUnordGreaterThanEqual [[#bool]] [[#r81]]
+; CHECK-SPIRV: 5 FUnordGreaterThanEqual [[#bool]] [[#r82]]
+; CHECK-SPIRV: 5 FUnordGreaterThanEqual [[#bool]] [[#r83]]
+; CHECK-SPIRV: 5 FUnordGreaterThanEqual [[#bool]] [[#r84]]
+; CHECK-SPIRV: 5 FUnordGreaterThanEqual [[#bool]] [[#r85]]
+; CHECK-SPIRV: 5 FUnordGreaterThanEqual [[#bool]] [[#r86]]
+; CHECK-SPIRV: 5 FUnordGreaterThanEqual [[#bool]] [[#r87]]
+; CHECK-SPIRV: 5 Unordered [[#bool]] [[#r88]]
+; CHECK-SPIRV: 5 Unordered [[#bool]] [[#r89]]
+; CHECK-SPIRV: 5 Unordered [[#bool]] [[#r90]]
+
+; CHECK-LLVM: %r1 = fcmp oeq float %a, %b
+; CHECK-LLVM: %r2 = fcmp nnan oeq float %a, %b
+; CHECK-LLVM: %r3 = fcmp ninf oeq float %a, %b
+; CHECK-LLVM: %r4 = fcmp nsz oeq float %a, %b
+; CHECK-LLVM: %r5 = fcmp arcp oeq float %a, %b
+; CHECK-LLVM: %r6 = fcmp fast oeq float %a, %b
+; CHECK-LLVM: %r7 = fcmp nnan ninf oeq float %a, %b
+; CHECK-LLVM: %r8 = fcmp one float %a, %b
+; CHECK-LLVM: %r9 = fcmp nnan one float %a, %b
+; CHECK-LLVM: %r10 = fcmp ninf one float %a, %b
+; CHECK-LLVM: %r11 = fcmp nsz one float %a, %b
+; CHECK-LLVM: %r12 = fcmp arcp one float %a, %b
+; CHECK-LLVM: %r13 = fcmp fast one float %a, %b
+; CHECK-LLVM: %r14 = fcmp nnan ninf one float %a, %b
+; CHECK-LLVM: %r15 = fcmp olt float %a, %b
+; CHECK-LLVM: %r16 = fcmp nnan olt float %a, %b
+; CHECK-LLVM: %r17 = fcmp ninf olt float %a, %b
+; CHECK-LLVM: %r18 = fcmp nsz olt float %a, %b
+; CHECK-LLVM: %r19 = fcmp arcp olt float %a, %b
+; CHECK-LLVM: %r20 = fcmp fast olt float %a, %b
+; CHECK-LLVM: %r21 = fcmp nnan ninf olt float %a, %b
+; CHECK-LLVM: %r22 = fcmp ogt float %a, %b
+; CHECK-LLVM: %r23 = fcmp nnan ogt float %a, %b
+; CHECK-LLVM: %r24 = fcmp ninf ogt float %a, %b
+; CHECK-LLVM: %r25 = fcmp nsz ogt float %a, %b
+; CHECK-LLVM: %r26 = fcmp arcp ogt float %a, %b
+; CHECK-LLVM: %r27 = fcmp fast ogt float %a, %b
+; CHECK-LLVM: %r28 = fcmp nnan ninf ogt float %a, %b
+; CHECK-LLVM: %r29 = fcmp ole float %a, %b
+; CHECK-LLVM: %r30 = fcmp nnan ole float %a, %b
+; CHECK-LLVM: %r31 = fcmp ninf ole float %a, %b
+; CHECK-LLVM: %r32 = fcmp nsz ole float %a, %b
+; CHECK-LLVM: %r33 = fcmp arcp ole float %a, %b
+; CHECK-LLVM: %r34 = fcmp fast ole float %a, %b
+; CHECK-LLVM: %r35 = fcmp nnan ninf ole float %a, %b
+; CHECK-LLVM: %r36 = fcmp oge float %a, %b
+; CHECK-LLVM: %r37 = fcmp nnan oge float %a, %b
+; CHECK-LLVM: %r38 = fcmp ninf oge float %a, %b
+; CHECK-LLVM: %r39 = fcmp nsz oge float %a, %b
+; CHECK-LLVM: %r40 = fcmp arcp oge float %a, %b
+; CHECK-LLVM: %r41 = fcmp fast oge float %a, %b
+; CHECK-LLVM: %r42 = fcmp nnan ninf oge float %a, %b
+; CHECK-LLVM: %r43 = call spir_func i32 @_Z9isorderedff(float %a, float %b)
```

Comment:
I don't know why it is translated as a function call. We have fcmp ord in LLVM IR.

Reply:
> I don't know why it is translated as a function call.

I guess the reason is that the translator is mainly built for OpenCL C, which has corresponding built-in function for that operation, which takes precedence over LLVM IR instruction

Reply:
What is the right behavior? To me translating to "fcmp ord" looks better. It is standard IR and can be recognized by LLVM compiler.

Reply:
I agree: to me, translating to the LLVM IR operation seems preferable. Exactly half of the floating-point relational and equality operations in SPIR-V have corresponding built-in functions in OpenCL C: for these, too, it is better that we translate to the LLVM IR operation, as we currently do, rather than to the OpenCL C built-in function.

Note that **OpFOrdEqual** and **OpFUnordNotEqual** correspond to **isequal()** and **isnotequal()** respectively, and that each of the relational **OpFOrd\*** operations has a corresponding relational built-in. There is no corresponding built-in function for **OpFUnordEqual**, **OpFOrdNotEqual** and the relational **OpFUnord\*** operations, although these do correspond to the negations of the relational builtins.

Would there be benefit in allowing fast-math decorations for **OpOrdered** or **OpUnordered**, also? Note that **OpLessOrGreater** is deprecated, so I do not believe it to be worth allowing fast-math decorations for that operation.

Note that **OpOrdered** and **OpUnordered** are specific to the **Kernel** execution mode. For other execution modes, they can be implemented easily using **OpIsNan** or by comparing each operand with itself.